feat(circuits)!: Assert clean public inputs on tail

sirasistant opened the pull request:
Resolves remaining TODOs from this PR https://github.com/AztecProtocol/aztec-packages/pull/18172

> Change the contract address of the protocol nullifier to NULL_MSG_SENDER_CONTRACT_ADDRESS, and the counter to 1, to satisfy the assumption of the sorting utils that counters of valid items are unique and non-zero.
TODO: this could be a problem if the first private function emits a nullifier of counter 1. Maybe we can enforce the start_side_effect_counter of the private calls to be non-zero (at least 1), then all side effects will have a counter > 1.

Approach taken: assert on init that the very first call of the tx starts with a counter > 1.

> In the tail circuits ([validate_accumulated_items.nr](https://github.com/AztecProtocol/aztec-packages/pull/18172/files#diff-29774a6210948cec2600f911db6fe523274b5dc005492b849a14f066085f86ef)), check one or some fields of an item (instead of whole_item.is_empty()) to determine nullish-ness.
TODO: See if we can only check the contract address for public call requests.
Use array lengths to compute gas used.
TODO: the gas used for l2-to-l1 messages does not consider the trimmed items (now only the recipient address is checked in the tail circuits), and they could be included in blobs even though they are nullish. We should change the array_length in tx_base_rollup to check the recipient address instead to get the number of non-nullish messages.

Approach taken: use an optimized checking approach so we can keep ensuring that items after the length are empty, but we just check one field for items within the length. This ensures that no garbage can exist after the array length propagated to the rollup, simplifying the interface.

LeilaWang commented on noir-projects/noir-protocol-circuits/crates/private-kernel-lib/src/components/private_call_data_validator.nr:
```diff
@@ -62,6 +62,12 @@ impl PrivateCallDataValidator {
 
+        // Any subsequent call must start with a side effect counter > than the init one
+        assert(
+            public_inputs.start_side_effect_counter > 1,
```
Maybe can add a comment here saying counter 1 is reserved for the protocol nullifier.

sirasistant replied:
sure!

LeilaWang reviewed (approved):
LGTM! 🎉

sirasistant committed:
feat(circuits)!: Assert clean public inputs on tail

Resolves remaining TODOs from this PR https://github.com/AztecProtocol/aztec-packages/pull/18172

> Change the contract address of the protocol nullifier to NULL_MSG_SENDER_CONTRACT_ADDRESS, and the counter to 1, to satisfy the assumption of the sorting utils that counters of valid items are unique and non-zero.
TODO: this could be a problem if the first private function emits a nullifier of counter 1. Maybe we can enforce the start_side_effect_counter of the private calls to be non-zero (at least 1), then all side effects will have a counter > 1.

Approach taken: assert on init that the very first call of the tx starts with a counter > 1.

> In the tail circuits ([validate_accumulated_items.nr](https://github.com/AztecProtocol/aztec-packages/pull/18172/files#diff-29774a6210948cec2600f911db6fe523274b5dc005492b849a14f066085f86ef)), check one or some fields of an item (instead of whole_item.is_empty()) to determine nullish-ness.
TODO: See if we can only check the contract address for public call requests.
Use array lengths to compute gas used.
TODO: the gas used for l2-to-l1 messages does not consider the trimmed items (now only the recipient address is checked in the tail circuits), and they could be included in blobs even though they are nullish. We should change the array_length in tx_base_rollup to check the recipient address instead to get the number of non-nullish messages.

Approach taken: use an optimized checking approach so we can keep ensuring that items after the length are empty, but we just check one field for items within the length. This ensures that no garbage can exist after the array length propagated to the rollup, simplifying the interface.
